src/loggers.jl

```diff
@@ -430,8 +430,18 @@ let
                     end
 
                     function $level(msg::Function, logger::Logger)
-                        log(msg, logger, $key)
-                        throw(ErrorException(msg()))
+                        # NOTE: We can't lazily evaluate do-blocks on errors or higher
+                        # because we need to determine whether the function returns
+                        # a string or an error type. We provide this syntax for readability
+                        # rather than performance.
+                        m = msg()
+                        if isa(m, String)
+                            log(logger, $key, m)
+                            throw(ErrorException(m))
+                        else
+                            log(logger, $key, sprint(io -> showerror(io, m)))
+                            throw(m)
+                        end
```

Comment:
I think you may want to check `m isa Exception` instead as someone may return a number or some other type which can be stringified.

Reply:
So you're thinking that we should fall back to string `string(x)`?

Reply:
Yes, mainly I was thinking about:
```julia
if isa(m, Exception)
    ...
else
    ...
end
```

Reply:
👌